qcodes_contrib_drivers/drivers/Attocube/AttocubeController.py

```diff
@@ -0,0 +1,324 @@
+from typing import List, Dict, Union, Optional, Sequence, Any, Tuple
+import numpy as np
+import visa
+import time
+import logging
+log = logging.getLogger(__name__)
+
+from qcodes.instrument.visa import VisaInstrument
+import qcodes.utils.validators as vals
+
+class AttocubeController(VisaInstrument):
+    """Base class for Attocube controller instrument.
+
+    This driver "low temperature" and "room temperature" modes, with different
+    voltage limits to protect the Attocubes.
+
+    I have found the Attocube serial communication to be kind of sketchy,
+    so ask_raw() is overridden try to make communication more robust. YMMV.
+
+    Args:
+        atto_config (dict): Configuration dict (example below).
+        temp (str): 'LT' or 'RT', depending on whether the attocubes are at room temp. or low temp.
+        kwargs: Keyword arguments to be passed to VisaInstrument constructor.
+
+    An example of the atto_config dictionary is provided below:
+    {
+        'name': 'atto',
+        'model': 'ANC300',
+        "address": "ASRL1::INSTR",
+        'timeout': 5,
+        'terminator': '\r\n',
+        'baud_rate' : 38400,
+        'axes': {'x': 1, 'y': 2, 'z': 3}, # {axis_name: controller_output}
+        'voltage_limits': {
+            'RT': {'x': 25, 'y': 25, 'z': 25}, # room temperature limits, V
+            'LT': {'x': 60, 'y': 60, 'z': 60} # low temperature limits, V
+        },
+        'default_frequency': {'x': 100, 'y': 100, 'z': 100}, # stepping frequency, Hz
+    }
+    """
+    def __init__(self, atto_config: Dict, temp: str, **kwargs) -> None:
+        super().__init__(atto_config['name'], atto_config['address'], atto_config['timeout'],
+                         atto_config['terminator'], **kwargs)
+        if temp.upper() not in ['LT', 'RT']:
+            raise ValueError('temp must be "LT" or "RT".')
+        if 'history' not in atto_config:
+            atto_config['history'] = {} # dict of past attocube movements, so you don't get lost
+        self.metadata.update(atto_config)
+        self.visa_handle.baud_rate = atto_config['baud_rate']
+        self.visa_handle.stop_bits = visa.constants.StopBits.one
+        self.visa_handle.parity = visa.constants.Parity.none
+        self.visa_handle.read_termination = '\r\n'
+        _ = self.parameters.pop('IDN') # Get rid of this parameter
+        self.temp = temp
+        self.timestamp_fmt = r'%Y-%m-%d_%H:%M:%S'
+        self.axes = self.metadata['axes']
+        self.voltage_limits = {}
+        self.add_parameter('version',
+                            label='Version',
+                            get_cmd='ver',
+                            get_parser=str,
+                            snapshot_get=False
+                            )
+        for axis, idx in self.axes.items():
+            self.voltage_limits.update({axis: self.metadata['voltage_limits'][temp][axis]}) # Volts
+            # axis mode
+            self.add_parameter('mode_ax{}'.format(idx),
```

Comment:
Instead of `.format()`, we prefer f-string notation when possible. Here, for example, instead of `"mode_ax{}".format(idx)`, preferred notation would be `f"mode_ax{idx}"`. This comment applies to the rest of the code.

Reply:
Okay, done here https://github.com/QCoDeS/Qcodes_contrib_drivers/pull/38/commits/9ec46d3f25aab9b62f236963b46b550ac71d2900